lib/puppet/parser/functions/defined_with_params.rb

```diff
@@ -24,7 +24,19 @@
     params = {}
   end
   ret = false
-  if resource = findresource(reference.to_s)
+
+  # Workaround for PE-20308
+  if reference.is_a?(String)
+    type_name, title = Puppet::Resource.type_and_title(reference, nil)
```

Comment:
The `Puppet::Resource#type_and_title` class method was introduced in Puppet 4.5.0. Perhaps that's what causing appveyor to fail for 4.2.3?